### DISCUSSION TITLE
[Proposal]: [Dot object deconstruction]

### DISCUSSION THREAD

**Discussion by fedeAlterio** [dotnet/csharplang#6198] [Language Ideas]
# Dot object deconstruction

* [x] Proposed
* [ ] Prototype: Not Started
* [ ] Implementation: Not Started
* [ ] Specification: Not Started

## Summary
[summary]: #summary

Allow creating a tuple from an object simply using the dot (.) operator.

## Motivation
The main motivation is lighter syntax and better readability.
Consider:

``` c#
var (from, to) = (request.LikeMessage.FromUsername, request.LikeMessage.ToUsername);
``` 

The idea is to pull the common part behind the dot (.).

``` c#
var (from, to) = request.LikeMessage.(FromUsername, ToUsername);
```

## Detailed design
Basically is a syntactic sugar:
``` c#
var (from, to) = request.LikeMessage.(FromUsername, ToUsername);
``` 

should be lowered to:
``` c#
var tmp = request.LikeMessage; 
var (from, to) = (tmp.FromUsername, tmp.ToUsername);
```

It could also be extended to methods:
``` c#
var (from, to) = request.LikeMessage.(GetFromUsername(), GetToUsername());
``` 

## Similar ideas
Consider also the assigments:
``` c#
request.likeMessage.FromUsername = "From";
request.likeMessage.ToUsername = "To";
``` 
Pulling out the dot:
``` c#
request.likeMessage.(FromUsername, ToUsername) = ("From", "To");
``` 
or with a syntax similar to object initialization:
``` c#
request.likeMessage.
{
   FromUsername = "From",
   ToUsername = "To"
};

``` 

## Drawbacks
The main drawback is the similarity of the expression to the function call operator.

## Alternatives
An alternative is to change the syntax to:
``` c#
var (from, to) = request.LikeMessage.{FromUsername, ToUsername};
```

**Comment by 333fred** [comment]
Very related to https://github.com/dotnet/csharplang/issues/4082, which is how I would prefer to approach this kind of problem.

**Reply by HaloFour** [reply]
Example:

```cs
var { FromUsername: from, ToUsername: to } = request.LikeMessage;
```

Pattern matching counterpart (and legal today):

```cs
if (request.LikeMessage is { FromUsername: var from, ToUsername: var to }) {
    // ...
}
```

**Reply by fedeAlterio** [reply]
Good point. I don't think they necessarily exclude each other. For example my proposal is an expression rather than a statement. So this should be valid code: 
``` c#
(string from, string to) GetFromTo(Request request) => request.LikeMessage.(FromUsername, ToUsername);
```

**Comment by En3Tho** [comment]
Also, this type of thing can be expanded with  assignments: e.g.

```csharp
request.(X, Y) = 1, 2
```

**Reply by fedeAlterio** [reply]
Really good idea. Maybe with syntax similar to object initialization. So
``` c#
request.
{
   X = 1,
   Y = 2
};
```
